File: core/core-api/src/main/java/com/coffee/gu/payment/PaymentManager.java
package com.coffee.gu.payment;

import com.coffee.gu.PGConfirmResult;
import com.coffee.gu.order.Order;

import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;


@Component
public class PaymentManager {
    private final PaymentRepository paymentRepository;

    public PaymentManager(PaymentRepository paymentRepository){
        this.paymentRepository = paymentRepository;
    }

    public Long createPayment(Order order, PaymentDiscount paymentDiscount) {
        Payment payment = paymentRepository.save(Payment.create(order, paymentDiscount));
        return payment.getId();
    }

    @Transactional
    public void pay(Payment payment, PGConfirmResult result) {
        payment.success(
                result.paymentKey(),
                result.paymentMethod(),
                result.approveCode(),
                result.approvedAt()
        );

        paymentRepository.save(payment);
    }

    public Payment prepare(Payment payment) {
        payment.prepare();
        return paymentRepository.save(payment);
    }

    public Payment save(Payment payment) {
        return paymentRepository.save(payment);
    }
}

File: core/core-api/src/main/kotlin/com/coffee/gu/payment/PaymentManager.kt
package com.coffee.gu.payment

import com.coffee.gu.PGConfirmResult
import com.coffee.gu.enums.PaymentMethod
import com.coffee.gu.order.Order
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.time.OffsetDateTime

@Component
class PaymentManager(
    private val paymentRepository: PaymentRepository,
) {
    fun createPayment(order: Order, paymentDiscount: PaymentDiscount): Long {
        val payment = paymentRepository.save(Payment.create(order, paymentDiscount))
        return payment.id
    }

    @Transactional
    fun pay(payment: Payment, result: PGConfirmResult) {
        payment.success(
            result.paymentKey,
            result.paymentMethod ?: PaymentMethod.CARD,
            result.approveCode,
            result.approvedAt ?: OffsetDateTime.now()
        )
        paymentRepository.save(payment)
    }

    fun prepare(payment: Payment): Payment {
        payment.prepare()
        return paymentRepository.save(payment)
    }

    fun save(payment: Payment): Payment {
        return paymentRepository.save(payment)
    }
}
